### ISSUE TITLE
Log user into Azure DevOps with MSAL

### ISSUE THREAD

**Issue by NightOwlCoder** [AzureAD/microsoft-authentication-library-for-dotnet#1952]
**Which Version of MSAL are you using ?**
`Microsoft.Identity.Client v4.16.1`

**Platform**
`Xamarin Forms`, with targets for `iOS`, `Droid`, `UWP` and `macOS`

**What authentication flow has the issue?**
* Desktop / Mobile
    * [x] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

Other? - please describe;

**Is this a new or existing app?**
we are developing a new app to work together with `Azure DevOps`.

**Repro**
I'm searching high and low for the "right" way of:
1. logging in our users
2. validate that user belongs to certain groups
3. execute queries on behalf of that user

**Expected behavior**
A clear and concise description of what to do.

**Actual behavior**
A very hard to follow situation.
One article points to `ADAL`, where the landing page says one must use `MSAL`.
But MSAL samples do not talk to any DevOps pagees.

**Possible Solution**
Better samples, addressing "normal" people problems, like:
1. login to `Azure DevOps` using MS account
2. get logged users group (or validate he belongs to certain groups)
3. execute actions on behalf of user

**Additional context/ Logs / Screenshots**
I've downloaded and followed every sample I could get my hands on, to try to add to my head the terms, ideas, concepts, etc.
But even after that, it has proven really hard to get to a working sample.

If you guys feel like, I could work with some of you creating better samples....

**Comment by jmprieur** [comment]
Which article is it, @NightOwlCoder ?
Are you using the Azure DevOps SDK?  https://docs.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-6.0 ?

**Comment by NightOwlCoder** [comment]
Yes, I've found that one also, that confirms (https://docs.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-6.0#authenticate) the same as this one:

https://docs.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/authentication-guidance?view=azure-devops

When you go to the ADAL page (https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/active-directory-authentication-libraries) it tells you to use MSAL:
> Note
> Looking for the Azure AD v2.0 libraries (MSAL)? Checkout the MSAL library guide.


And then I get lost, because Azure Dev Ops has different `token` and `authorize` endpoints.

At this point I'm "almost" implementing it "by hand", but this looks like a so common scenario that I want to check with your guys first...

**Comment by bgavrilMS** [comment]
Hi @NightOwlCoder - ADAL library is officially [deprecated](https://techcommunity.microsoft.com/t5/azure-active-directory-identity/update-your-applications-to-use-microsoft-authentication-library/ba-p/1257363) and the replacement is MSAL. 

I logged an [issue ](https://github.com/microsoft/azure-devops-auth-samples/issues/56) on Azure DevOps sample to upgrade. 

I think it should be fairly easy to user MSAL to get a token for Azure DevOps. By looking in Azure Portal > AAD > App Registration > Api Permissions I can see that Azure DevOps exposes a single scope _https://app.vssps.visualstudio.com/user_impersonation_

Identity SDK maintains a ton of samples that show you how to get a token for a protected resource, including Xamarin samples (although if you are new to Identity SDK, I'd recommend you start with a simple WPF or console app and then move to Xamarin)

https://docs.microsoft.com/en-gb/azure/active-directory/develop/sample-v2-code

**Comment by jmprieur** [comment]
@NightOwlCoder the sample to look at is https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/tree/master/1.%20Desktop%20app%20calls%20Web%20API

just as @bgavrilMS said you'd replace the URL by the Azure Devops URL, and the scope by  https://app.vssps.visualstudio.com/user_impersonation (or  https://app.vssps.visualstudio.com/.default)
